### PR TITLE
[SPARK-25233][Streaming] Give the user the option of specifying a minimum message per partition per batch when using kafka direct API with backpressure

### PR DESCRIPTION
**Pull request by rezasafi**
After SPARK-18371, it is guaranteed that there would be at least one message per partition per batch using direct kafka API when new messages exist in the topics. This change will give the user the option of setting the minimum instead of just a hard coded 1 limit
The related unit test is updated and some internal tests verified that the topic partitions with new messages will be progressed by the specified minimum.

### CODE REVIEW

**Comment by rezasafi** [comment]
@koeninger @arzt will appreciate your feedback here. Thank you in advance.

**Comment by koeninger** [comment]
Jenkins, ok to test

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala [hunk 1]
```diff
@@ -141,10 +143,9 @@ private[spark] class DirectKafkaInputDStream[K, V](
           tp -> Math.max(offset - currentOffsets(tp), 0)
         }
         val totalLag = lagPerPartition.values.sum
-
         lagPerPartition.map { case (tp, lag) =>
           val maxRateLimitPerPartition = ppc.maxRatePerPartition(tp)
-          val backpressureRate = lag / totalLag.toDouble * rate
+          var backpressureRate = lag / totalLag.toDouble * rate
```
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala]: Why was this changed to a var?

**Review by koeninger — CHANGES_REQUESTED** [review]

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala (hunk 1 above)
Comment (rezasafi) [reply external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala]: Sorry that was a mistake from another change I was trying to do. Forgot to revert. will fix this

File: docs/configuration.md [hunk 2]
```diff
@@ -1925,6 +1925,14 @@ showDF(properties, numRows = 200, truncate = FALSE)
     first batch when the backpressure mechanism is enabled.
   </td>
 </tr>
+<tr>
+  <td><code>spark.streaming.backpressure.fixedMinMessagePerPartition</code></td>
```
[the remark this answers is not in the document]
Comment (rezasafi) [reply docs/configuration.md]: Yeah that make sense. Will apply your comment. Thanks

**Review by koeninger — APPROVED** [review]
LGTM pending tests passing after fixing that whitespace style issue

Will leave it open for a while in case any others want to comment, if you don't hear back by next week feel free to ping me again for a potential merge

**Comment by rezasafi** [comment]
Jenkins, retest this please

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala [hunk 3]
```diff
@@ -154,7 +153,8 @@ private[spark] class DirectKafkaInputDStream[K, V](
     if (effectiveRateLimitPerPartition.values.sum > 0) {
       val secsPerBatch = context.graph.batchDuration.milliseconds.toDouble / 1000
       Some(effectiveRateLimitPerPartition.map {
-        case (tp, limit) => tp -> Math.max((secsPerBatch * limit).toLong, 1L)
+        case (tp, limit) => tp -> Math.max((secsPerBatch * limit).toLong,
+          Math.max(ppc.minRatePerPartition(tp), 1L))
```
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala]: Is the second Math.max actually necessary?
The default implementation of minRatePerPartition will be 1 anyway.
If someone makes a custom implementation that e.g. returns zero, should they get what they asked for?.

Comment (rezasafi) [reply external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala]: I just didn't want to break the reasoning behind SPARK-18371 to have at least 1 always. I didn't have any other reason for this. I can change it to give the user the freedom.

**Comment by koeninger** [comment]
I think since the default behavior is still 1, it's probably ok to let
someone do what they want here

**Comment by rezasafi** [comment]
Thank you very much @koeninger
